packages/gatsby-plugin-image/src/components/gatsby-image.browser.tsx

```diff
@@ -50,171 +53,233 @@ export interface IGatsbyImageData {
+      `[data-gatsby-image-ssr]`
     )
+    // On first server hydration do nothing
+    if (hasNativeLazyLoadSupport() && hasSSRHtml && !this.hydrated.current) {
+      this.hydrated.current = true
+      return Promise.resolve()
+    }
+
+    return import(`./lazy-hydrate`).then(({ lazyHydrate }) => {
+      this.lazyHydrator = lazyHydrate(
+        {
+          image: props.image.images,
+          isLoading: state.isLoading,
+          isLoaded: state.isLoaded,
+          toggleIsLoaded: () => {
+            props.onLoad?.()
+
+            this.setState({
+              isLoaded: true,
+            })
+          },
+          ref: this.ref,
+          ...props,
+        },
+        this.root,
+        this.hydrated
+      )
+    })
   }
 
-  const { style: wStyle, className: wClass, ...wrapperProps } = getWrapperProps(
-    width,
-    height,
-    layout
-  )
+  /**
+   * Choose if setupIntersectionObserver should use the image cache or not.
+   */
+  _setupIntersectionObserver(useCache = true): void {
+    import(`./intersection-observer`).then(({ createIntersectionObserver }) => {
+      const intersectionObserver = createIntersectionObserver(() => {
+        if (this.root.current) {
+          const cacheKey = JSON.stringify(this.props.image.images)
+          this.props.onStartLoad?.({
+            wasCached: useCache && hasImageLoaded(cacheKey),
+          })
+          this.setState({
+            isLoading: true,
+            isLoaded: useCache && hasImageLoaded(cacheKey),
+          })
+        }
+      })
+
+      if (this.root.current) {
+        // @ts-ignore - hello
+        this.unobserveRef = intersectionObserver(this.root)
+      }
+    })
+  }
+
+  shouldComponentUpdate(nextProps, nextState): boolean {
+    let hasChanged = false
+
+    // this check mostly means people do not have the correct ref checks in place, we want to reset some state to suppport loading effects
+    if (this.props.image.images !== nextProps.image.images) {
+      // reset state, we'll rely on intersection observer to reload
+      if (this.unobserveRef) {
+        // unregister intersectionObserver
+        this.unobserveRef(this.root)
+
+        // // on unmount, make sure we cleanup
+        if (this.hydrated.current && this.lazyHydrator) {
+          render(null, this.root.current)
+        }
+      }
+
+      this.setState(
+        {
+          isLoading: false,
+          isLoaded: false,
+        },
+        () => {
+          this._setupIntersectionObserver(false)
+        }
+      )
+
+      hasChanged = true
+    }
+
+    if (this.root.current && !hasChanged) {
+      console.log(nextState)
+      this._lazyHydrate(nextProps, nextState)
+    }
+
+    return false
+  }
 
-  useEffect((): (() => void) | undefined => {
-    if (root.current) {
-      const hasSSRHtml = root.current.querySelector(
+  componentDidMount(): void {
+    if (this.root.current) {
+      const hasSSRHtml = this.root.current.querySelector(
```

Comment:
This is consfusing variable naming. `hasSSRHtml` implies it's a boolean, but it's actually the element itself, and we access its properties later. We should change it to something like `ssrElement`